x-pack/plugins/security_solution/server/lib/detection_engine/index/delete_all_index.ts

```diff
@@ -4,14 +4,42 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { IndicesDeleteParams } from 'elasticsearch';
+import { IndicesDeleteParams, Client } from 'elasticsearch';
 import { CallWithRequest } from '../types';
 
+const MAX_ATTEMPTS = 5;
+
 export const deleteAllIndex = async (
-  callWithRequest: CallWithRequest<IndicesDeleteParams, boolean>,
-  index: string
+  callWithRequest: CallWithRequest<IndicesDeleteParams, ReturnType<Client['indices']['getAlias']>>,
+  pattern: string
 ): Promise<boolean> => {
-  return callWithRequest('indices.delete', {
-    index,
-  });
+  for (let attempt = 1; ; attempt++) {
+    if (attempt > MAX_ATTEMPTS) {
+      throw new Error(
+        `Failed to delete indexes with pattern [${pattern}] after ${MAX_ATTEMPTS} attempts`
+      );
+    }
+
+    // resolve pattern to concrete index names
+    const resp = await callWithRequest('indices.getAlias', {
+      index: pattern,
+      ignore: 404,
+    });
+
+    if (resp.status === 404) {
+      return true;
+    }
+
+    const indices = Object.keys(resp) as string[];
+
+    // if no indexes exits then we're done with this pattern
+    if (!indices.length) {
+      return true;
+    }
+
+    // delete the concrete indexes we found and try again until this pattern resolves to no indexes
+    await callWithRequest('indices.delete', {
+      index: indices,
+    });
```

Comment:
Is it possible to add a `refresh: wait_for` here for us since we're looping now? It should be ok for us and our users and not a big breaking change. Might make tests and other things less in-deterministic and get us a better chance of getting this right on the first loop around.

Reply:
The index delete API doesn't support refresh, and AFAIK is a "synchronous" action and the entire cluster will refuse to search across shards for deleted indexes.

Reply:
> "synchronous" action and the entire cluster will refuse to search across shards for deleted indexes immediately

Do you mean, when we execute a request like `DELETE /delme-idx-1,delme-idx-2`, before returning `"acknowledged" : true` every node in the cluster must acknowledge the deletion and disable all other actions with the indices - synchronously? To me "acknowledged" semantically means "ok, received your request, will process it" w/o time guarantees, so I wouldn't personally rely on that. But I might be wrong.

I think my question here is: after deleting indices in `await callWithRequest('indices.delete'`, can we still get the same indices from `await callWithRequest('indices.getAlias'` in the next iteration? If the cluster is under load or whatever? Can we catch an error due to a race condition? Because if it's possible, on the next iteration when trying to delete the same index that was already asked to be deleted, we can receive an error.

I tried this:

```
PUT delme-idx-1/_doc/1
{ "foo": "bar" }

DELETE /delme-idx-1,delme-idx-2
```

where deletion returns an error

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "index_not_found_exception",
        "reason" : "no such index [delme-idx-2]",
        "index_uuid" : "_na_",
        "resource.type" : "index_or_alias",
        "resource.id" : "delme-idx-2",
        "index" : "delme-idx-2"
      }
    ],
    "type" : "index_not_found_exception",
    "reason" : "no such index [delme-idx-2]",
    "index_uuid" : "_na_",
    "resource.type" : "index_or_alias",
    "resource.id" : "delme-idx-2",
    "index" : "delme-idx-2"
  },
  "status" : 404
}
```

and then

```
GET /delme-idx-*/_alias
```

returns

```
{
  "delme-idx-1" : {
    "aliases" : { }
  }
}
```

Which means that `DELETE /delme-idx-1,delme-idx-2` did nothing.

Is a similar situation possible due to a race condition?

Reply:
And just thinking out loud.. Maybe specifying a timeout + exp wait would do the job?

```
namesToDelete = get index names from the pattern
send the delete request

while (not timed out) {
  existingNames = get index names
  if (namesToDelete not in existingNames) {
    return
  }
  get next delay // exponential
  wait
}

throw timed out
```

Here the difference is: this will delete only those specific indices which existed at the moment of receiving the request. So if a new index matching the pattern is added during the deletion (while loop), it will be ignored. Which can be good or bad depending on the case.

Reply:
@banderror You definitely have a point. I am pretty darn confident that a successful response from Elasticsearch means that the cluster state has been updated and all nodes have acknowledged the update, that said there isn't any reason to not protect ourselves from a situation like this. To do so I've updated the `indices.delete` call to use the `ignore_unavailable=true` query string param which will ignore concrete indices which are deleted, preventing the 404 you showed and will instead just delete the indices that do exist.

Reply:
WRT exponential backoff/timeout I'm not totally opposed, but I also feel like it's a good deal more complicated to implement and I don't really see the value personally. If you specify a wildcard to this API and new indexes show up that match the wildcard while it's running I think it makes sense to delete them too. Additionally, if new indexes keep showing up leading to a maximum number of iterations that's a issue which I don't think this API is intended to solve. Finally, picking a timeout seems like a very challenging task as operations to ES can take quite some time in certain scenarios and I think we should rely on the configured request timeout rather than defining a new timeout for this operation specifically.

Reply:
`ignore_unavailable=true` sounds good, thank you! 👍